File: Week4/VolvoCodingChallenge/app/src/main/java/com/yeah/ruisu/volvocodingchallenge/data/RepositoryModule/RepositoryImpl.java
package com.yeah.ruisu.volvocodingchallenge.data.RepositoryModule;

import android.annotation.SuppressLint;
import android.arch.lifecycle.MutableLiveData;

import com.yeah.ruisu.volvocodingchallenge.data.remote.RemoteServiceHelper;
import com.yeah.ruisu.volvocodingchallenge.data.remote.models.ForecastDatum;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class RepositoryImpl implements Repository
{
    private MutableLiveData<List<ForecastDatum>> weatherForecastLiveData = new MutableLiveData<>();
    private RemoteServiceHelper MyRemoteServiceHelper;

    public RepositoryImpl(RemoteServiceHelper InRemoteServiceHelper)
    {
        this.MyRemoteServiceHelper = InRemoteServiceHelper;
    }

    @SuppressLint("CheckResult")
    @Override
    public void getWeatherForecast(String WOEID, String year, String month, String day)
    {
        MyRemoteServiceHelper.getWeatherForecast(WOEID, year, month, day)
                                .subscribeOn(Schedulers.io())
                                .observeOn(AndroidSchedulers.mainThread())
                                .subscribe(weatherApiResponse ->
                                        weatherForecastLiveData.setValue(weatherApiResponse),
                                        Throwable::printStackTrace);
    }

    public MutableLiveData<List<ForecastDatum>> getWeatherForecastLiveData() {
        return weatherForecastLiveData;
    }
}

File: week5/RealVolvoCodingChallenge/app/src/main/java/com/yeah/ruisu/realvolvocodingchallenge/data/RepositoryModule/RepositoryImpl.kt
package com.yeah.ruisu.realvolvocodingchallenge.data.RepositoryModule

import android.arch.lifecycle.MutableLiveData
import android.util.Log
import com.yeah.ruisu.realvolvocodingchallenge.data.remote.RemoteServiceHelper
import com.yeah.ruisu.realvolvocodingchallenge.data.remote.models.ForecastDatum
import kotlinx.coroutines.experimental.CoroutineExceptionHandler
import kotlinx.coroutines.experimental.launch
import retrofit2.HttpException

class RepositoryImpl (private val remoteServiceHelper: RemoteServiceHelper) : Repository
{
    override val weatherForecastLiveData: MutableLiveData<List<ForecastDatum>> = MutableLiveData()

    override fun getWeatherForecast(WOEID: String)
    {
        val exceptionHandler = CoroutineExceptionHandler {_,
                                                         throwable ->
            if (throwable is HttpException)
            {
                Log.d("main", "In HttpException")
            }
            else
            {
                Log.d("Main", "Other error", throwable.fillInStackTrace())
            }
        }

        launch (exceptionHandler)
        {
            val metaWeatherResponse = remoteServiceHelper.getWeatherForecast(WOEID)
                                                            .await()
            weatherForecastLiveData.postValue(metaWeatherResponse)
        }
    }
}